cmd/deploy-helm/helm.go

```diff
@@ -36,140 +37,123 @@ type helmChart struct {
 }
 
 // commonHelmUpgradeArgs returns arguments common to "helm upgrade" and "helm diff upgrade".
-func commonHelmUpgradeArgs(
-	releaseName, helmArchive string,
-	opts options,
-	valuesFiles, cliValues []string,
-	targetConfig *config.Environment,
-) ([]string, error) {
-	args, err := shlex.Split(opts.upgradeFlags)
+func (d *deployHelm) commonHelmUpgradeArgs() ([]string, error) {
+	args, err := shlex.Split(d.opts.upgradeFlags)
 	if err != nil {
-		return []string{}, fmt.Errorf("parse upgrade flags (%s): %s", opts.upgradeFlags, err)
+		return []string{}, fmt.Errorf("parse upgrade flags (%s): %s", d.opts.upgradeFlags, err)
 	}
-	if opts.debug {
+	if d.opts.debug {
 		args = append([]string{"--debug"}, args...)
 	}
-	if targetConfig.APIServer != "" {
+	if d.targetConfig.APIServer != "" {
 		args = append(
 			[]string{
-				fmt.Sprintf("--kube-apiserver=%s", targetConfig.APIServer),
-				fmt.Sprintf("--kube-token=%s", targetConfig.APIToken),
+				fmt.Sprintf("--kube-apiserver=%s", d.targetConfig.APIServer),
+				fmt.Sprintf("--kube-token=%s", d.targetConfig.APIToken),
 			},
 			args...,
 		)
 	}
-	for _, vf := range valuesFiles {
+	for _, vf := range d.valuesFiles {
 		args = append(args, fmt.Sprintf("--values=%s", vf))
 	}
-	args = append(args, cliValues...)
-	args = append(args, releaseName, helmArchive)
+	args = append(args, d.cliValues...)
+	args = append(args, d.releaseName, d.helmArchive)
 	return args, nil
 }
 
-// helmUpgrade runs given Helm command.
-func helmUpgrade(args []string, stdout, stderr io.Writer) error {
-	return command.Run(
-		helmBin, args, []string{fmt.Sprintf("SOPS_AGE_KEY_FILE=%s", ageKeyFilePath)},
-		stdout, stderr,
+// getHelmChart reads given filename into a helmChart struct.
+func getHelmChart(filename string) (*helmChart, error) {
+	y, err := os.ReadFile(filename)
+	if err != nil {
+		return nil, fmt.Errorf("read chart file: %w", err)
+	}
+
+	var hc *helmChart
+	err = yaml.Unmarshal(y, &hc)
+	if err != nil {
+		return nil, fmt.Errorf("unmarshal chart: %w", err)
+	}
+	return hc, nil
+}
+
+// getChartVersion extracts the version from given Helm chart.
+func getChartVersion(contextVersion string, hc *helmChart) string {
+	if len(contextVersion) > 0 && contextVersion != pipelinectxt.WIP {
+		return contextVersion
+	}
+	return hc.Version
+}
+
+// cleanHelmDiffOutput removes error messages from the given Helm output.
+// Those error messages are confusing, because they do not come from  an actual
+// error, but from detecting drift between desired and current Helm state.
+func cleanHelmDiffOutput(out string) string {
+	if !strings.Contains(out, helmDiffDetectedMarker) {
+		return out
+	}
+	cleanedOut := strings.Replace(
+		out, helmDiffDetectedMarker, desiredDiffMessage, -1,
 	)
+	r := regexp.MustCompile(`Error: plugin "(diff|secrets)" exited with error[\n]?`)
+	cleanedOut = r.ReplaceAllString(cleanedOut, "")
+	r = regexp.MustCompile(`helm.go:81: \[debug\] plugin "(diff|secrets)" exited with error[\n]?`)
```

Comment:
what if a new helm version changes the line where they log that error from? suggestion: to not depend on binary not line values

Reply:
Yeah this is very far from ideal. That said, this regex hasn't changed in my refactoring. There is a test that ensures this regex actually works https://github.com/opendevstack/ods-pipeline/blob/master/test/tasks/ods-deploy-helm_test.go#L94. So if a new version changes this, we should know.